src/OrleansRuntime/Streams/QueueBalancer/QueueBalancerBaseClass.cs

```diff
@@ -0,0 +1,72 @@
+﻿using Orleans.Providers;
+using Orleans.Runtime;
+using Orleans.Streams;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// Base class for StreamQueueBalancer
+    /// </summary>
+    public abstract class QueueBalancerBaseClass : IStreamQueueBalancer
+    {
+        protected readonly List<IStreamQueueBalanceListener> queueBalanceListeners;
+        public QueueBalancerBaseClass()
+        {
+            this.queueBalanceListeners = new List<IStreamQueueBalanceListener>();
+        }
+        public abstract IEnumerable<QueueId> GetMyQueues();
+        public abstract Task Initialize(string strProviderName, IStreamQueueMapper queueMapper, TimeSpan siloMaturityPeriod, IProviderConfiguration providerConfig);
+        public virtual bool SubscribeToQueueDistributionChangeEvents(IStreamQueueBalanceListener observer)
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException("observer");
+            }
+            lock (queueBalanceListeners)
+            {
+                if (queueBalanceListeners.Contains(observer))
+                {
+                    return false;
+                }
+                queueBalanceListeners.Add(observer);
+                return true;
+            }
+        }
+        public virtual bool UnSubscribeFromQueueDistributionChangeEvents(IStreamQueueBalanceListener observer)
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException("observer");
+            }
+            lock (queueBalanceListeners)
+            {
+                return queueBalanceListeners.Contains(observer) && queueBalanceListeners.Remove(observer);
+            }
+        }
+
+
+        public static List<string> GetActiveSilos(ISiloStatusOracle siloStatusOracle, ConcurrentDictionary<SiloAddress, bool> immatureSilos)
```

Comment:
If you look at its child class, such as LeaseBasedQueueBalancer. You will have more clue.
Basically, whenever there's a SiloStatusNotification, code calls siloStatusOracle asking for active silo information. If there's new silo which was recorded in immatureSilos, then the code would wait `siloMaturityPeriod `to mark it as mature. If before it is marked as mature, GetActiveSilos is called, and siloStatusOracle report those new silo as active, we still think they are immature. 

I think it is a conservative way to deal with inconsistent view in a distributed cluster. Since this local siloStatusOracle thinks this silo is active doesn't mean other silo thinks it is active too. So we'd better wait for `siloMaturityPeriod`. 

It requires a concurrentDic because there could be under multi-thread scenario. For example, siloStatusOracle trigger SiloStatusChangeNotification will be running in one thread, and pullingAgentManager will be triggering GetMyQueues in another thread, both of which will need to accees immatureSIlos